src/state_tracker/state.rs

```diff
@@ -66,77 +66,66 @@ where
         S: From<&'a [u8]>,
     {
         use self::{State::*, Token::*};
-
-        match (self.state.pop(), *token) {
+        let last_index = self.state.len().max(1) - 1;
+        let actual_len = self.state.len();
+        match (self.state.last_mut(), *token) {
             (None, End) => {
                 return self.latch_err(Err(E::from(StructureError::invalid_state(
                     "End not allowed at top level",
                 ))));
             },
-            (Some(Seq), End) => {},
-            (Some(MapKey(_)), End) => {},
+            (Some(Seq), End) | (Some(MapKey(_)), End) => {
+                self.state.pop();
+            },
             (Some(MapKey(None)), String(label)) => {
-                self.state.push(MapValue(S::from(label)));
+                self.state[last_index] = MapValue(S::from(label)); //TODO: looks similar!
             },
-            (Some(MapKey(Some(oldlabel))), String(label)) => {
-                if oldlabel.as_ref() >= label {
-                    return self.latch_err(Err(E::from(StructureError::UnsortedKeys)));
-                }
-                self.state.push(MapValue(S::from(label)));
+            (Some(MapKey(Some(oldlabel))), String(label)) if oldlabel.as_ref() >= label => {
+                self.state.pop();
+                return self.latch_err(Err(E::from(StructureError::UnsortedKeys)));
+            },
+            (Some(MapKey(Some(_oldlabel))), String(label)) => {
+                self.state[last_index] = MapValue(S::from(label));
             },
-            (Some(oldstate @ MapKey(_)), _tok) => {
-                self.state.push(oldstate);
+            (Some(_oldstate @ MapKey(_)), _tok) => {
+                self.state.pop();
```

Comment:
I'm not sure that this pop is actually necessary, and it's technically a behaviour change. However, it shouldn't matter because we immediately enter an error state anyways, at which point the state becomes irrelevant.